projects/interfaces/buffer.go

```diff
@@ -0,0 +1,42 @@
+package interfaces
+
+import (
+	"slices"
+	"errors"
+)
+
+//OurByteBuffer - Limitations
+// - the buffer grows and shrinks as data is added and read from it. The maximum data stored by the data is dependent on the machine in which the code is run. The buffer frees up all memory that has been successfully read from the buffer.
```

Comment:
> I wouldn't ordinarily have brought this up, but as you explicitly stated this, I'll point you at https://go.dev/blog/slices-intro#a-possible-gotcha - have a read about when memory is actually freed up when using slices and see if you see anything that contradicts what this comment says :)

I found it - `Since the slice references the original array, as long as the slice is kept around the garbage collector can’t release the array`
In the case of OurByteBuffer all I did was to delete (reslicing and returning the modified slice) the bytes that are just read. The returned slice which I set as the new buffer still references the original array so memory is not freed.